cip/CIP2016-06-14-Define-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,377 @@
+
+====== New map equality ======
+
+To rectify this, we propose instead that two maps `m1` and `m2` should be equal if:
+
+* `m1` and `m2` have the same keys,
+** including keys that map to a `null` value (the order of keys as returned by `keys()` does not matter here).
+* Additionally, for each such key `k`,
+** `m1.k = m2.k` is `true`.
+
+As a consequence of these changes, plain <<equality-def>> is not reflexive for all values (consider: `{a: null} = {a: null}`, `[null] = [null]`). However this was already the case (consider: `null = null` \=> `null`).
+
+Note that <<equality-def>> is reflexive for values that do not involve `null` though.
+
+[[orderability-equivalence]]
+=== Orderability and equivalence ===
+
+We propose that <<orderability-def>> and <<equivalence-def>> should be aligned with each other, i.e.
+
+`expr1` is equivalent to `expr2` if and only if they have the same position under orderability (i.e. they would be sorted before (or after resp.) any other non-equivalent value in the same way).
+
+Orderability and equivalence produce <<missing-null,"missing" `null` values>>.
+
+[[orderability-def,orderability]]
+==== Orderability ====
+
+[[global-sort-order,global sort order]]
+We propose that orderability be defined between any pair of values, where the result is always `true` or `false`.
+
+To accomplish this, there must be a pre-determined order of types and each value must fall under exactly one disjoint leaf type in this order. We propose to use the following ascending global sort order of disjoint types:
+
+* `MAP` types
+** <<regular-map,Regular map>>
+** `NODE`
+** `RELATIONSHIP`
+* `LIST OF ANY?`
+* `PATH`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER`
+** `NaN` is treated as the largest number in orderability only (i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
+
+To give a concrete example, under this global sort order all nodes come before all strings.
+
+Between values of the same type in the global sort order, orderability defers to comparability except that equality is overriden by equivalence as described below.
+
+The accompanying descending global sort order is the same order in reverse (i.e. it runs from `VOID` to `MAP`).
+
+[[equivalence-def,equivalence]]
+==== Equivalence ====
+
+Equivalence remains mostly unchanged but now can be defined succinctly as being identical to equality except that:
+
+* Any two `null` values are equivalent (both directly or inside nested structures).
+* Any two `NaN` values are equivalent (both directly or inside nested structures).
+* However, `null` and `NaN` values are not equivalent (both directly or inside nested structures).
+
+Equivalence is reflexive for all values.
+
+[[aggregation]]
+=== Aggregation ===
+
+Generally, an aggregation function call `aggr(expr)` processes all matching rows for each aggregation key found in an incoming record (keys are compared using equivalence, as noted above).
```

Comment:
I don't think of aggregation as "function calls", could we not just call it "aggregation", rather than "aggregation function call"?